### ISSUE TITLE
Tensorflow 2 creating custom dataset 

### ISSUE THREAD

**Issue by ArashJavan** [tensorflow/tensorflow#31540]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 2.0.0-beta
- Are you willing to contribute it (Yes/No): yes

**Describe the feature and the current behavior/state.**
IMO there is currently not a really clean way to build an own dataset by using/subclassing the tf.data.Dataset-class. Specially for datasets which have more complex structure as [ICDAR](https://rrc.cvc.uab.es/?ch=4), here the labels are made of text-files with bounding-boxes and transcripts. In these cases even tensorflow_datasets can not be used in a straight way as described [here] (https://www.tensorflow.org/datasets/add_dataset).

So i would like to know what is actually the correct procedure to use the Dataset-Api for creating a custom dataset? 

**Will this change the current api? How?**

**Who will benefit with this feature?**
Most of object detection tasks

**Any Other info.**

**Comment by braindotai** [comment]
This is a very common issue I facing every time I have to create a custom data, and this always makes be __switch back to pytorch__.

Last time I needed to work with celeb-a dataset, and tensorflow_dataset is giving me some `has wrong checksum` issue, so I had absolutely no choice but to manually download and create a dataset... but guess what, celeb a is not a type of dataset where each image belongs to a single category, therefore its totally impossible to read this image folder as dataset in tensorflow.

And ofc, if I'd have tried appending all the images and their labels in an array to use `.from_tensor_slices` then it would blow the memory.

Its such a big and silly disappointment :(

**Comment by braindotai** [comment]
> Have you taken a look at [TensorFlow Datasets](https://www.tensorflow.org/datasets)? It offers support to users who would like to add a [custom dataset](https://www.tensorflow.org/datasets/add_dataset), rather than just use `tf.data.Dataset`.

`tensorflow_datasets` has nothing to do with creating a custom dataset, yes it does provide most of the datasets out of the box but when we make a product we most of the time never those datasets. 

Adding a custom dataset, again has nothing to do with creating a custom one with sub classing.

Please have a look [here](https://pytorch.org/tutorials/beginner/data_loading_tutorial.html) and tell if any of this can be done in tensorflow?

**Comment by braindotai** [comment]
@arashwan Here's an example how you can create a fully custom dataset using `tf.data.Dataset`. I'm reading a celeb_a dataset, that is located at `'dataset'` directory. Here's how the dataset directory looks like...
```
dataset
├── img_align_celeba
│   └── img_align_celeba
├── list_attr_celeba.csv
├── list_bbox_celeba.csv
├── list_eval_partition.csv
└── list_landmarks_align_celeba.csv
```

```python
image_paths = sorted(glob.glob(os.path.join('dataset', 'img_align_celeba', 'img_align_celeba', '*.jpg')))

df = pd.read_csv('dataset/list_attr_celeba.csv')
df.replace(to_replace = -1, value = 0, inplace = True)
labels = df.iloc[:, 1:].values

print(image_paths[:2])
print(labels[:2])
# prints
'''
['dataset/img_align_celeba/img_align_celeba/000001.jpg', 'dataset/img_align_celeba/img_align_celeba/000002.jpg']
[[0 1 1 0 0 0 0 0 0 0 0 1 0 0 0 0 0 0 1 1 0 1 0 0 1 0 0 1 0 0 0 1 1 0 1 0
  1 0 0 1]
 [0 0 0 1 0 0 0 1 0 0 0 1 0 0 0 0 0 0 0 1 0 1 0 0 1 0 0 0 0 0 0 1 0 0 0 0
  0 0 0 1]]
'''

# here's the tensorflow part...
@tf.function
def read_image(image_path):
    image = tf.io.read_file(image_path)
    image = tf.image.decode_image(image, channels = 3, dtype = tf.float32)
    return image

@tf.function
def normalize(image):
    image = (image - tf.reduce_min(image))/(tf.reduce_max(image) - tf.reduce_min(image))
    image = (2 * image) - 1
    return image

@tf.function
def augment(image):
    image = tf.image.random_crop(image, (178, 178, 3))
    image = tf.image.resize(image, (256, 256))
    image = tf.image.random_flip_left_right(image)
    image = tf.image.random_saturation(image, 0.5, 2.0)
    image = tf.image.random_brightness(image, 0.5)
    return image

@tf.function
def preprocess(image_path, label):
    image = read_image(image_path)
    image = augment(image)
    image = normalize(image)
    return image, label

dataset = tf.data.Dataset.from_tensor_slices((image_paths, labels))
dataset = dataset.map(preprocess) # you can do almost all sorts of custom preprocesses....
dataset = dataset.shuffle(buffer_size = 1024)
dataset = dataset.batch(batch_size = 128)
dataset = dataset.prefetch(buffer_size = 128)

for x, y in dataset:
    break

print(x.shape, y.shape)
# prints
# (128, 256, 256, 3) (128, 40)
```

Hope this helps

**Comment by ArashJavan** [comment]
@braindotai and @dynamicwebpaige thanks for your suggestions I will try them and report ASAP.
